platform/wayland/cog-platform-wl.c

```diff
@@ -1968,24 +2188,18 @@ init_wayland (GError **error)
 }
 
 static void
-clear_wayland (void)
+clear_wayland(void)
 {
     g_clear_pointer(&wl_data.event_src, g_source_destroy);
+    win_data.shell_functions->destroy_shell();
```

Comment:
Perhaps it's worth clearing the `shell_functions` pointer after calling `destroy_shell` - this would make the situation where a shell function is erroneously called after this function easier to debug.